File: ros_ws/src/navegation/src/control.py
#!/usr/bin/env python

#ROS dependencies
import rospy
from std_msgs.msg import Float32
from custom_msgs.msg import vel_cmd
from geometry_msgs.msg import Pose2D, PoseStamped

#Import numpy
import numpy as np
import math

q  = np.array([[0, 0, 0]]).T
qd = np.array([[0, 0, 0]]).T

#parameters
r = 0.05 #wheel radius
d = 0.08 #distance between wheels
h = 0.02 #distance between center and new point

k = np.array([[1, 0],
              [0, 2]])
cmd = vel_cmd()

def controlCalc():
    global q, qd, r, d, h, k, cmd, control, left_pub, right_pub

    qe = q[0:2] - qd[0:2]

        #u = np.array([[1,5]]).T
    D = np.array([[r/2*np.cos(q[2,0]) - h*r/d*np.sin(q[2,0]), r/2*np.cos(q[2,0]) + h*r/d*np.sin(q[2,0])],
                  [r/2*np.sin(q[2,0]) + h*r/d*np.cos(q[2,0]), r/2*np.sin(q[2,0]) - h*r/d*np.cos(q[2,0])]])

    u = np.dot(np.linalg.inv(D),(np.dot(-k,qe)))

    cmd.wr.data = u[0, 0]
    cmd.wl.data = u[1, 0]

    control.publish(cmd)

    left_pub.publish(cmd.wl)
    right_pub.publish(cmd.wr)    

def newEstimationCallback(pose_estimation):
    global q

    #convert from quaternion to euler
    # save orientation
    w = pose_estimation.pose.orientation.w 
    x = pose_estimation.pose.orientation.x
    y = pose_estimation.pose.orientation.y 
    z = pose_estimation.pose.orientation.z 

    #apply convertion
    t1 = +2.0 * (w * z + x * y)
    t2 = +1.0 - 2.0 * (y * y + z * z)
    yaw = math.atan2(t1, t2)

    #save states
    q[0, 0] = pose_estimation.pose.position.x
    q[1, 0] = pose_estimation.pose.position.y 
    q[2, 0] = yaw
    controlCalc()

def newGoalCallback(goal_pose):
    global qd
    qd[0,0] = goal_pose.x
    qd[1,0] = goal_pose.y
    qd[2,0] = goal_pose.theta

def main():

    global control, left_pub, right_pub

    rospy.init_node("Control")
	
    l_speed = rospy.get_param("/navigation/topics/cmd_vel_l", "/WOMBAT/navegation/cmd_l")
    r_speed = rospy.get_param("/navigation/topics/cmd_vel_r", "/WOMBAT/navegation/cmd_r")

    control  = rospy.Publisher("/WOMBAT/navegation/control", vel_cmd, queue_size=1)
    left_pub = rospy.Publisher(l_speed, Float32, queue_size=1)
    right_pub = rospy.Publisher(r_speed, Float32, queue_size=1)

    pose_sub = rospy.Subscriber("/WOMBAT/navegation/pose", PoseStamped, newEstimationCallback, queue_size = 10)

    goal_sub = rospy.Subscriber("/WOMBAT/navegation/goal", Pose2D, newGoalCallback, queue_size = 1)
    rospy.spin()

if __name__ == "__main__":
    main()

File: ros_ws/src/navegation/src/control.cpp
#include "headers/control.h"

void control_shutdown(int sig)
{
    delete control::ctr_h;
    
    // Stop the motors
    control::right_msg.data = 0;
    control::left_msg.data  = 0;

    control::right_pub.publish(control::right_msg);
    control::left_pub.publish(control::left_msg);

    ros::shutdown();
}

void pose_callback(const geometry_msgs::Pose2D::ConstPtr& pose)
{
    // Read values from Pose2D message
    Eigen::Matrix<double, 3, 1> q; 
    q << pose->x,
         pose->y,
         pose->theta;

    // Get the time difference
    double dt = ros_utils::dt_and_swp(control::t, ros::Time::now());

    // Make control calculations
    Eigen::Matrix<double, 2, 1> u = control::ctr_h->control_position(q, control::qd, dt);

    // Set values to messages and publish them
    control::right_msg.data = u(0,0);
    control::left_msg.data  = u(1,0);

    control::right_pub.publish(control::right_msg);
    control::left_pub.publish(control::left_msg);
}

void goal_callback(const geometry_msgs::Pose2D::ConstPtr& goal)
{
    control::qd << goal->x,
                   goal->y;
}

int main(int argc, char** argv)
{
    // Start ROS node and node handle
    ros::init(argc, argv, "control");
    ros::NodeHandle nh;
    
    // Set shutdown function
    signal(SIGINT, control_shutdown);

    // Get ROS params
    std::string left_topic;
    nh.param<std::string>("/navigation/topics/vel_l", left_topic, "/WOMBAT/navegation/leftSpeed");

    std::string right_topic;
    nh.param<std::string>("/navigation/topics/vel_r", right_topic, "/WOMBAT/navegation/rightSpeed");

    double r, d, h;

    nh.param<double>("/navigation/model_args/radius", r, 0.05);

    nh.param<double>("/navigation/model_args/distance", d, 0.08);

    nh.param<double>("/navigation/model_args/displacement", h, 0.02);
    
    // Asign constant control values
    double k1, k2;
    k1 = 0.1;
    k2 = 0.2;

    control::ctr_h = new Control(k1, k2, r, d, h);

    // Create ROS subscribers
    
    //WOMBAT/navegation/pose2d
    ros::Subscriber pose_sub = nh.subscribe("/slam_out_pose", 10, pose_callback);
    ros::Subscriber goal_sub = nh.subscribe("/WOMBAT/navegation/goal", 10, goal_callback);

    // Create ROS publishers
    control::left_pub = nh.advertise<std_msgs::Float32>(left_topic, 10);
    control::right_pub = nh.advertise<std_msgs::Float32>(right_topic, 10);

    // Initialize t and qd
    control::qd << 0,
                   0;
    control::t = ros::Time::now();

    // Call ROS spin
    ros::spin();

    return 0;
}

    
